Update AOT for .NET 8 (#36397)
dotnet/docs: docs/core/deploying/native-aot/index.md

The native AOT deployment model uses an ahead-of-time compiler to compile IL to native code at the time of publish. Native AOT apps don't use a just-in-time (JIT) compiler when the application runs. Native AOT apps can run in restricted environments where a JIT isn't allowed. Native AOT applications target a specific runtime environment, such as Linux x64 or Windows x64, just like publishing a [self-contained app](../index.md#publish-self-contained).

## Limitations in the .NET native AOT deployment model

### [.NET 7](#tab/net7)

> > [!WARNING]

Reviewer: The warning here is a bit weird. Console apps were fully supported with NativeAOT in .NET 7 - but this styling looks like it's something to avoid.

Author: #36529

> > 

Author: ```suggestion
>
```

> There are some limitations in the .NET native AOT deployment model, with the main one being that run-time code generation isn't possible. For more information, see [Limitations of Native AOT deployment](#limitations-of-native-aot-deployment). The support in the .NET 7 release is targeted towards console-type applications.

Author: > Should this be folded under the .NET 7 tab just below as well?

Folded it into the .NET 7 tab, great suggestion.

## AOT-compatibility analyzers

### [.NET 7](#tab/net7)

> AOT-compatibility analyzers are available only in .NET 8 and later versions.

Author: See #36516 which logs this

### [.NET 8+](#tab/net8plus)

By default, native AOT publishing produces debug information in a separate file:

> - Linux: *.dbg* 

Author: ```suggestion
- Linux: *.dbg*
```

> - Windows: *.pdb* 

Author: ```suggestion
- Windows: *.pdb*
```

> -  *.pdb* on Windows

Author: ```suggestion
- *.pdb* on Windows
```

> - Linux: *.dbg* 
> - Windows: *.pdb* 
> - macOS *.dwarf* 

Author: ```suggestion
- Linux: *.dbg* 
- Windows: *.pdb* 
- macOS: *.dwarf* 
```

> - macOS: *.dwarf* 

Author: ```suggestion
- macOS: *.dwarf*
```

> - *.dbg* on Linux
> - *.pdb* on Windows
> - *.dwarf* on macOS

Author: ```suggestion
- Linux: *.dbg* 
- Windows: *.pdb* 
- macOS *.dwarf* 
```

> The debug is necessary for running the app under the debugger or inspecting crash dumps. On Unix-like platforms, set the `StripSymbols` property to `false` to include the debug information in the native binary. Including debug information makes the native binary larger.

Author: ```suggestion
The debug file is necessary for running the app under the debugger or inspecting crash dumps. On Unix-like platforms, set the `StripSymbols` property to `false` to include the debug information in the native binary. Including debug information makes the native binary larger.
```

### [.NET 8+](#tab/net8plus)

> By default, native AOT publishing produces debug information in a separate file on all supported platforms (*.dbg* on Linux, *.pdb* on Windows, and *.dwarf* on macOS). This file is necessary for running the app under the debugger or inspecting crash dumps. On Unix-like platforms, set the `StripSymbols` property to `false` to include the debug information in the native binary. Including the debug information makes the native binary larger.

Author: ```suggestion
By default, native AOT publishing produces debug information in a separate file:

- *.dbg* on Linux
-  *.pdb* on Windows
- *.dwarf* on macOS

The debug is necessary for running the app under the debugger or inspecting crash dumps. On Unix-like platforms, set the `StripSymbols` property to `false` to include the debug information in the native binary. Including debug information makes the native binary larger.
```

## Limitations of native AOT deployment

Native AOT apps have the following limitations:

- No dynamic loading, for example, `Assembly.LoadFile`.
- No run-time code generation, for example, `System.Reflection.Emit`.
- No C++/CLI.
- Windows: No built-in COM.
- Requires trimming, which has [limitations](../trimming/incompatibilities.md).
- Implies compilation into a single file, which has known [incompatibilities](../single-file/overview.md#api-incompatibility).
- Apps include required runtime libraries (just like [self-contained apps](../index.md#publish-self-contained), increasing their size as compared to framework-dependent apps).
- <xref:System.Linq.Expressions> always use their interpreted form, which is slower than run-time generated compiled code.
- Not all the runtime libraries are fully annotated to be native AOT compatible. That is, some warnings in the runtime libraries aren't actionable by end developers.

Part of the publish process:

- Analyzes the entire project and its dependencies.
- Produces warnings whenever the limitations of the published app might be encountered at run time.

### OS specific limitations

### [.NET 7](#tab/net7)

> Should be targeted for console type apps. ASP.NET Core is ***not*** supported.
> + Limited diagnostic support for debugging and profiling.

Author: ```suggestion
- Should be targeted for console type apps. ASP.NET Core is ***not*** supported.
- Limited diagnostic support for debugging and profiling.
```

## Limitations of native AOT deployment

Native AOT apps have the following limitations:

- No dynamic loading, for example, `Assembly.LoadFile`.
- No run-time code generation, for example, `System.Reflection.Emit`.
- No C++/CLI.
- Windows: No built-in COM.
- Requires trimming, which has [limitations](../trimming/incompatibilities.md).
- Implies compilation into a single file, which has known [incompatibilities](../single-file/overview.md#api-incompatibility).
- Apps include required runtime libraries (just like [self-contained apps](../index.md#publish-self-contained), increasing their size as compared to framework-dependent apps).
- <xref:System.Linq.Expressions> always use their interpreted form, which is slower than run-time generated compiled code.
- Not all the runtime libraries are fully annotated to be native AOT compatible. That is, some warnings in the runtime libraries aren't actionable by end developers.

> - Limited diagnostic support for debugging and profiling.

Author: ```suggestion
```

Part of the publish process:

- Analyzes the entire project and its dependencies.
- Produces warnings whenever the limitations of the published app might be encountered at run time.

### OS specific limitations

### [.NET 7](#tab/net7)

Should be targeted for console type apps. ASP.NET Core is ***not*** supported.

> 

Author: ```suggestion
+ Limited diagnostic support for debugging and profiling.
```

## Limitations of native AOT deployment

Native AOT apps have the following limitations:

- No dynamic loading, for example, `Assembly.LoadFile`.
- No run-time code generation, for example, `System.Reflection.Emit`.
- No C++/CLI.
- Windows: No built-in COM.
- Requires trimming, which has [limitations](../trimming/incompatibilities.md).
- Implies compilation into a single file, which has known [incompatibilities](../single-file/overview.md#api-incompatibility).
- Apps include required runtime libraries (just like [self-contained apps](../index.md#publish-self-contained), increasing their size as compared to framework-dependent apps).
- <xref:System.Linq.Expressions> always use their interpreted form, which is slower than run-time generated compiled code.

> - Should be targeted for console type apps, ASP.NET Core is not supported.

Author: ```suggestion
```

## Limitations of native AOT deployment

### [.NET 7](#tab/net7)

Native AOT apps have limitations and compatibility issues:

- No dynamic loading, for example, `Assembly.LoadFile`.
- No run-time code generation, for example, `System.Reflection.Emit`.
- No C++/CLI.
- Windows: No built-in COM.
- Requires trimming, which has [limitations](../trimming/incompatibilities.md).
- Implies compilation into a single file, which has known [incompatibilities](../single-file/overview.md#api-incompatibility).
- Apps include required runtime libraries (just like [self-contained apps](../index.md#publish-self-contained), increasing their size as compared to framework-dependent apps).
- <xref:System.Linq.Expressions> always use their interpreted form, which is slower than run-time generated compiled code.
- Should be targeted for console type apps, ASP.NET Core is not supported.
- Not all the runtime libraries are fully annotated to be native AOT compatible. That is, some warnings in the runtime libraries aren't actionable by end developers.
- Limited diagnostic support for debugging and profiling.

> The publish process:

Reviewer: I find this change problematic. Before this it reads like "one part of the publish process analyzes the entire project ...". After the change it reads like "the publish process is these two steps - analysis and producing warnings", which is wrong and misleading.

I think the original intent of this sentence was to point out that there are warning produced by the publish process which will alert the developer if most of the above limitations are encountered by the app. It was not here to describe what the publish process does (lot of things).

Reviewer: Applies to the .net 8 version as well - same story.

Author: > I find this change problematic. Before this it reads like "one part of the publish process analyzes the entire project ...". After the change it reads like "the publish process is these two steps - analysis and producing warnings", which is wrong and misleading.
> 
> I think the original intent of this sentence was to point out that there are warning produced by the publish process which will alert the developer if most of the above limitations are encountered by the app. It was not here to describe what the publish process does (lot of things).

Can you suggest a rewriting of the original. The sentence is way too long and I was splitting it up.

Reviewer: Maybe something like:
"The publish process will detect if running the app may encounter some of the limitations and will produce warnings accordingly". (not a native speaker, so my grammar is terrible)

> The publish process:

Author: ```suggestion
Part of the publish process:
```

- Analyzes the entire project and its dependencies.
- Produces warnings whenever the limitations of the published app might be encountered at run time.

### [.NET 8+](#tab/net8plus)

> Native AOT apps have limitations and compatibility issues:

Reviewer 2: In the .NET 8 list I am wondering if it would make sense to explicitly call out that we support ASP.NET minimal APIs with native AOT with a few limitations?

Author: > In the .NET 8 list I am wondering if it would make sense to explicitly call out that we support ASP.NET minimal APIs with native AOT with a few limitations?

Good suggestion, I'll add that and link off.

Reviewer 3: However, not all of ASP.NET is supported. Just minimal APIs and gRPC. MVC, Razor Pages, Blazor, and SignalR are all unsupported.

Reviewer 4: We should just link to https://learn.microsoft.com/en-us/aspnet/core/fundamentals/native-aot?view=aspnetcore-8.0

- No dynamic loading, for example, `Assembly.LoadFile`.
- No run-time code generation, for example, `System.Reflection.Emit`.
- No C++/CLI.
- Windows: No built-in COM.
- Requires trimming, which has [limitations](../trimming/incompatibilities.md).

> - Implies compilation into a single file, which has known [incompatibilities](../single-file/overview.md#api-incompatibility).

Reviewer 3: Would it help to refactor the parts that are always going to be limitations of Native AOT apps so they aren't duplicated for every release?

Reviewer: +1 to this - I don't see us fixing most of these probably ever.

- Apps include required runtime libraries (just like [self-contained apps](../index.md#publish-self-contained), increasing their size as compared to framework-dependent apps).
- <xref:System.Linq.Expressions> always use their interpreted form, which is slower than run-time generated compiled code.
- Not all the runtime libraries are fully annotated to be native AOT compatible. That is, some warnings in the runtime libraries aren't actionable by end developers.

> - Limited diagnostic support for debugging and profiling.

Reviewer 3: Diagnostic support has been addressed in .NET 8. We have way better diagnostic support in .NET 8. I'm not sure what is left as a limitation.

cc @LakshanF @agocke @tommcdon

Reviewer: There are still quite a few limitations and UX differences. Lakshan is working on a doc about the diag story for NativeAOT on .NET 8. Once that is done we should just like to that from here.

Author: > There are still quite a few limitations and UX differences. Lakshan is working on a doc about the diag story for NativeAOT on .NET 8. Once that is done we should just ~like~ link to that from here.

#36502

Can we get this approved so I can S&M?

## Limitations of native AOT deployment

Native AOT apps have the following limitations:

- No dynamic loading, for example, `Assembly.LoadFile`.
- No run-time code generation, for example, `System.Reflection.Emit`.
- No C++/CLI.
- Windows: No built-in COM.
- Requires trimming, which has [limitations](../trimming/incompatibilities.md).
- Implies compilation into a single file, which has known [incompatibilities](../single-file/overview.md#api-incompatibility).
- Apps include required runtime libraries (just like [self-contained apps](../index.md#publish-self-contained), increasing their size as compared to framework-dependent apps).
- <xref:System.Linq.Expressions> always use their interpreted form, which is slower than run-time generated compiled code.
- Not all the runtime libraries are fully annotated to be native AOT compatible. That is, some warnings in the runtime libraries aren't actionable by end developers.

> The publish process detects if the running the app may encounter some of the limitations. If publishing detects potential limitations, warnings are displayed.

Author: ```suggestion
The publish process analyzes the entire project and its dependencies for possible limitations. Warnings are issued for each limitation the published app may encountered at run time.
```

## Limitations of native AOT deployment

Native AOT apps have the following limitations:

- No dynamic loading, for example, `Assembly.LoadFile`.
- No run-time code generation, for example, `System.Reflection.Emit`.
- No C++/CLI.
- Windows: No built-in COM.
- Requires trimming, which has [limitations](../trimming/incompatibilities.md).
- Implies compilation into a single file, which has known [incompatibilities](../single-file/overview.md#api-incompatibility).
- Apps include required runtime libraries (just like [self-contained apps](../index.md#publish-self-contained), increasing their size as compared to framework-dependent apps).
- <xref:System.Linq.Expressions> always use their interpreted form, which is slower than run-time generated compiled code.
- Not all the runtime libraries are fully annotated to be native AOT compatible. That is, some warnings in the runtime libraries aren't actionable by end developers.

The publish process analyzes the entire project and its dependencies for possible limitations. Warnings are issued for each limitation the published app may encountered at run time.

> ### OS specific limitations

Reviewer 3: Is this section supposed to be empty?

Author: fixed

> ### OS specific limitations

Author: #36529

> ### [.NET 7](#tab/net7)

Author: ```suggestion
### [.NET 7+](#tab/net7plus)
```

- Should be targeted for console type apps. ASP.NET Core is ***not*** supported.
- Limited diagnostic support for debugging and profiling.

### [.NET 8+](#tab/net8plus)

> Support for some ASP.NET Core features. For more information, see [ASP.NET Core support for native AOT](/aspnet/core/fundamentals/native-aot/?view=aspnetcore-8.0&preserve-view=true).

Reviewer: The diagnostics limitations are still in place for .NET 8 - some of them were fixed, but I would not call it done at all.

Author: #36529

The following table shows supported compilation targets.

### [.NET 7](#tab/net7)

| Platform | Supported architecture |
|----------|------------------------|
| Windows  | x64, Arm64             |
| Linux    | x64, Arm64             |

### [.NET 8+](#tab/net8plus)

| Platform | Supported architecture |
|----------|------------------------|
| Windows  | x64, Arm64             |
| Linux    | x64, Arm64             |

> | macOS in | x64, Arm64      |

Author: ```suggestion
| macOS | x64, Arm64      |
```

> | macOS    | x64, Arm64             |

Author: > @ivanpovazan @kotlarmilos @akoeplinger @steveisok How would you like to go about iOS/tvOS/etc. support? Should we document it here with an asterisk that it's experimental, or omit for now?

That only shows up on the .NET 8 tab. I removed the `*` because it's easy to miss. This makes it obvious.

> | macOS    | x64, Arm64             |

Reviewer 5: > @ivanpovazan @kotlarmilos @akoeplinger @steveisok How would you like to go about iOS/tvOS/etc. support? Should we document it here with an asterisk that it's experimental, or omit for now?

I would vote +1 to add it with an asterisk (noting it is experimental).
Speaking of which, what about linux bionic?

The following table shows supported compilation targets.

### [.NET 7](#tab/net7)

| Platform | Supported architecture |
|----------|------------------------|
| Windows  | x64, Arm64             |
| Linux    | x64, Arm64             |

### [.NET 8+](#tab/net8plus)

| Platform | Supported architecture |
|----------|------------------------|
| Windows  | x64, Arm64             |
| Linux    | x64, Arm64             |

> | macOS    | x64, Arm64             |

Author: ```suggestion
| Platform | Supported architecture |  Notes               |
|----------|------------------------|----------------------|
| Windows  | x64, Arm64             |                      |
| Linux    | x64, Arm64             |                      |
| ma

> | Platform | Supported architecture |  Notes               |
> |----------|------------------------|----------------------|
> | Windows  | x64, Arm64             |                      |
> | Linux    | x64, Arm64             |                      |
> | ma

Author: ```suggestion
| Platform | Supported architecture |  Notes               |
|----------|------------------------|----------------------|
| Windows  | x64, Arm64             |                      |
| Linux    | x64, Arm64             |                      |
| macOS    | x64, Arm64             | Experimental support |
```

The following table shows supported compilation targets.

### [.NET 7](#tab/net7)

| Platform | Supported architecture |
|----------|------------------------|
| Windows  | x64, Arm64             |
| Linux    | x64, Arm64             |

### [.NET 8+](#tab/net8plus)

| Platform | Supported architecture |  Notes               |
|----------|------------------------|----------------------|
| Windows  | x64, Arm64             |                      |
| Linux    | x64, Arm64             |                      |

> | macOS    | x64, Arm64             | Experimental support |

Reviewer 6: ```suggestion
| macOS   | x64, Arm64 |                                        |
| iOS     | x64, Arm64 | Experimental support                   |
| Android | x64, Arm64 | Experimental, no built-in Java interop |
```

@ivanpovazan what's the full matrix here? Should tvOS/watchOS/Catalyst also be called out (I don't actually know what we're bringing up :))?

The following table shows supported compilation targets.

### [.NET 7](#tab/net7)

| Platform | Supported architecture |
|----------|------------------------|
| Windows  | x64, Arm64             |
| Linux    | x64, Arm64             |

### [.NET 8+](#tab/net8plus)

| Platform | Supported architecture |  Notes               |
|----------|------------------------|----------------------|
| Windows  | x64, Arm64             |                      |
| Linux    | x64, Arm64             |                      |

> | macOS   | x64, Arm64 |                                        |
> | iOS     | x64, Arm64 | Experimental support                   |
> | Android | x64, Arm64 | Experimental, no built-in Java interop |

Reviewer 5: ```suggestion
| macOS   | x64, Arm64 |                                        |
| iOS     | Arm64 | Experimental support                   |
| iOSSimulator     | x64, Arm64 | Experimental support                   |
| tvOS     | Arm64 | Experimental support                   |
| tvOSSimulator     | x64, Arm64 | Experimental support                   |
| MacCatalyst     | x64, Arm64 | Experimental support                   |
| Android | x64, Arm64 | Experimental, no built-in Java interop |
```

@MichalStrehovsky
